contracts/mainnet/connectors/notional/main.sol

```diff
@@ -0,0 +1,671 @@
+            ? uint256(_assetCashChange)
+            : 0;
+
+        setUint(setId, assetCashChange);
+
+        _eventName = "LogRedeemNTokenRaw(address,uint16,bool,uint96,int256)";
+        _eventParam = abi.encode(
+            address(this),
+            currencyId,
+            sellTokenAssets,
+            tokensToRedeem,
+            assetCashChange
+        );
+    }
+
+    /**
+     * @notice Redeems nTokens to cash and withdraws the resulting cash
+     * @dev Also possible to use redeemNTokenRaw and withdrawCollateral to achieve the same
+     * result but this is more gas efficient, it does it in one call to Notional
+     * @param currencyId notional defined currency id of nToken
+     * @param tokensToRedeem amount of nTokens to redeem
+     * @param amountToWithdraw amount of asset cash to withdraw, if set to uint(-1) then will withdraw the
+     * entire cash balance in notional
+     * @param redeemToUnderlying if true, will redeem the asset cash withdrawn to underlying tokens
+     * @param getId id of amount of tokens to redeem
+     * @param setId id to set amount of asset cash or underlying tokens withdrawn
+     */
+    function redeemNTokenAndWithdraw(
+        uint16 currencyId,
+        uint96 tokensToRedeem,
+        uint256 amountToWithdraw,
+        bool redeemToUnderlying,
+        uint256 getId,
+        uint256 setId
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        tokensToRedeem = getNTokenRedeemAmount(
+            currencyId,
+            tokensToRedeem,
+            getId
+        );
+
+        BalanceAction[] memory action = new BalanceAction[](1);
+        action[0].actionType = DepositActionType.RedeemNToken;
+        action[0].currencyId = currencyId;
+        action[0].depositActionAmount = tokensToRedeem;
+        action[0].redeemToUnderlying = redeemToUnderlying;
+        if (amountToWithdraw == uint256(-1)) {
+            // This setting will override the withdrawAmountInternalPrecision
+            action[0].withdrawEntireCashBalance = true;
+        } else {
+            action[0].withdrawAmountInternalPrecision = amountToWithdraw;
+        }
+
+        executeActionWithBalanceChange(
+            action,
+            0,
+            currencyId,
+            redeemToUnderlying,
+            setId
+        );
+
+        _eventName = "LogRedeemNTokenWithdraw(address,uint16,uint96,uint256,bool)";
+        _eventParam = abi.encode(
+            address(this),
+            currencyId,
+            tokensToRedeem,
+            amountToWithdraw,
+            redeemToUnderlying
+        );
+    }
+
+    /**
+     * @notice Redeems nTokens and uses the cash to repay a borrow.
+     * @dev When specifying fCashAmount be sure to calculate it such that the account
+     * has enough cash after redeeming nTokens to pay down the debt. This can be done
+     * off-chain using the Notional SDK.
```

Comment:
Can we do this on-chain inside contract itself?

Reply:
The calculation is not straightforward due to slippage across multiple markets, it's not easy to implement this is in solidity. Note that this method is here for convenience, a user could achieve the same results in two steps without using the off-chain SDK (first redeem, then repay).

Reply:
Alright, got it